File: p151_175/p157/src/Main.java

public class Main {

	public static long count(int n) {
		long count = 0L;
		long power10 = 1L;
		for (int nIndex = 0; nIndex < n; nIndex++) {
			power10 *= 10L;
		}
		for (long a = 1L; a <= 2 * power10; a++) {
			for (long p = power10/a + 1; p <= 2 * power10 / a; p++) {
				long bDenominator = a * p - power10;
				if ((a * power10) % bDenominator == 0) {
					count ++;
				}
			}
		}
		return count;
	}
	
	/**
	 * @param args
	 */
	public static void main(String[] args) {
		long total = 0L;
		for(int n = 1; n <= 9; n++) {
			long count = count(n);
			System.out.printf("%d\t%d\n", n, count);
			total += count;
		}
		System.out.println(total);
	}
}

File: p151_175/p157/main.rb
# b = (a 10^n)/(ap - 10^n)

def count(n)
  count = 0
  power_10 = 10 ** n
  (1..(2 * power_10)).each do |a|
    puts "#{100 * a/(2 * power_10)}%" if a % (10 ** 7) == 0
    ((power_10/a + 1)..(2 * power_10/a)).each do |pp|
      b_denominator = a * pp - power_10
      next if (a * power_10) % b_denominator != 0
      b = (a * power_10) / b_denominator
      count += 1
      # p [a, b, pp]
    end
  end
  count
end

total = 0
(1..6).each do |n|
  count = count(n)
  p [n, count]
  total += count
end
p total
